### ISSUE TITLE
batch_normlization op的实现上的疑惑？

### ISSUE THREAD

**Issue by Dely-Yu** [PaddlePaddle/Paddle#19640]
我对batch-norm 有一些不清楚。 https://github.com/PaddlePaddle/Paddle/blob/release/1.5/python/paddle/fluid/layers/nn.py#L3049

官方源码里面： mean，variance；mean_out，variance_out； saved_mean，saved_variance分别指什么呢？ 

按照源码：mean，variance是全局的，在一个batch中，对mean，variance更新，得到新的mean_out，variance_out。saved_mean，saved_variance 是用在inference的。 

我的疑问是 ：

1.  按照batch_norm的传统，inference的时候应该用全局的mean，variance。  

2. saved_mean，saved_varianc 与 mean，variance有什么不同呢？
[图片]
[图片][图片]

**Comment by Dely-Yu** [comment]
当batchsize 为10的时候，我把mean_out 和 saved_mean打印出来，发现数值上很接近，但是相差10倍。
1567660471	mean_out	The place is:CPUPlace
Tensor[Block_5_layer2_bn1.w_1]
	shape: [512,]
	dtype: f
	data: 6.25614e-07,6.05691e-07,3.71493e-07,4.25677e-07,-8.28404e-08,-4.7798e-07,8.96869e-07,2.77358e-07,-9.39467e-07,-4.27785e-07,1.75467e-08,7.98441e-08,-9.49192e-07,-3.50926e-07,3.98922e-07,7.76319e-08,-9.2629e-08,6.75401e-07,-6.60171e-08,-9.6001e-08,5.99713e-07,-5.95415e-08,3.09607e-08,-1.12361e-07,4.87125e-07,7.41937e-07

1567660471	saved_mean	The place is:CPUPlace
Tensor[Block_5_layer2_bn1.tmp_0]
	shape: [512,]
	dtype: f
	data: 6.25614e-06,6.05691e-06,3.71493e-06,4.25677e-06,-8.28403e-07,-4.7798e-06,8.96869e-06,2.77358e-06,-9.39467e-06,-4.27785e-06,1.75467e-07,7.98441e-07,-9.49191e-06,-3.50926e-06,3.98922e-06,7.76319e-07,-9.26289e-07,6.75401e-06,-6.60171e-07,-9.6001e-07,5.99713e-06,-5.95415e-07,3.09606e-07,-1.12361e-06,4.87125e-06,7.41937e-06,

**Comment by Dely-Yu** [comment]
储存batchnorm层的参数：batch_norm_0.b_0  batch_norm_0.w_0  batch_norm_0.w_1  batch_norm_0.w_2

分别对应了代码里的：bias，scale，mean，variance。 saved_mean的名字是Block_5_layer2_bn1.tmp_0，不是储存的参数。 那么推断时载入的参数？

**Comment by Superjomn** [comment]
<img width="512" alt="image" src="https://user-images.githubusercontent.com/328693/64316037-40e9c280-cfe6-11e9-85bb-cda06896b8e9.png">

这个图片可以上传下 @Dely-Yu

**Comment by ceci3** [comment]
你好，mean和variance指的是全局的均值方差，saved_mean saved_variance指的是当前batch的均值方差。infer的时候用的是mean和variance。

**Comment by Dely-Yu** [comment]
![14DD5648085574406BCB836A36C766B2](https://user-images.githubusercontent.com/29305857/64316264-e8ff8b80-cfe6-11e9-9e9b-09aecb83c528.JPG)
![AB10DB1C633AA603B4E7289926B79662](https://user-images.githubusercontent.com/29305857/64316266-ebfa7c00-cfe6-11e9-85ab-88dab30f786e.JPG)
![3238C7CBF9F944DD02A4DA5D7F6C887F](https://user-images.githubusercontent.com/29305857/64316271-eef56c80-cfe6-11e9-92fa-d1c66df5fa64.JPG)

**Comment by ceci3** [comment]
> 当batchsize 为10的时候，我把mean_out 和 saved_mean打印出来，发现数值上很接近，但是相差10倍。
> 1567660471	mean_out	The place is:CPUPlace
> Tensor[Block_5_layer2_bn1.w_1]
> shape: [512,]
> dtype: f
> data: 6.25614e-07,6.05691e-07,3.71493e-07,4.25677e-07,-8.28404e-08,-4.7798e-07,8.96869e-07,2.77358e-07,-9.39467e-07,-4.27785e-07,1.75467e-08,7.98441e-08,-9.49192e-07,-3.50926e-07,3.98922e-07,7.76319e-08,-9.2629e-08,6.75401e-07,-6.60171e-08,-9.6001e-08,5.99713e-07,-5.95415e-08,3.09607e-08,-1.12361e-07,4.87125e-07,7.41937e-07
> 
> 1567660471	saved_mean	The place is:CPUPlace
> Tensor[Block_5_layer2_bn1.tmp_0]
> shape: [512,]
> dtype: f
> data: 6.25614e-06,6.05691e-06,3.71493e-06,4.25677e-06,-8.28403e-07,-4.7798e-06,8.96869e-06,2.77358e-06,-9.39467e-06,-4.27785e-06,1.75467e-07,7.98441e-07,-9.49191e-06,-3.50926e-06,3.98922e-06,7.76319e-07,-9.26289e-07,6.75401e-06,-6.60171e-07,-9.6001e-07,5.99713e-06,-5.95415e-07,3.09606e-07,-1.12361e-06,4.87125e-06,7.41937e-06,

这是因为全局均值会对当前均值做平滑操作：mean = momentum * mean + (1.0 - momentum) * saved_mean，batch norm里默认的momentum是0.9， 所以出现你看到的相差10倍的情况

**Comment by Dely-Yu** [comment]
> > 当batchsize 为10的时候，我把mean_out 和 saved_mean打印出来，发现数值上很接近，但是相差10倍。
> > 1567660471	mean_out	The place is:CPUPlace
> > Tensor[Block_5_layer2_bn1.w_1]
> > shape: [512,]
> > dtype: f
> > data: 6.25614e-07,6.05691e-07,3.71493e-07,4.25677e-07,-8.28404e-08,-4.7798e-07,8.96869e-07,2.77358e-07,-9.39467e-07,-4.27785e-07,1.75467e-08,7.98441e-08,-9.49192e-07,-3.50926e-07,3.98922e-07,7.76319e-08,-9.2629e-08,6.75401e-07,-6.60171e-08,-9.6001e-08,5.99713e-07,-5.95415e-08,3.09607e-08,-1.12361e-07,4.87125e-07,7.41937e-07
> > 1567660471	saved_mean	The place is:CPUPlace
> > Tensor[Block_5_layer2_bn1.tmp_0]
> > shape: [512,]
> > dtype: f
> > data: 6.25614e-06,6.05691e-06,3.71493e-06,4.25677e-06,-8.28403e-07,-4.7798e-06,8.96869e-06,2.77358e-06,-9.39467e-06,-4.27785e-06,1.75467e-07,7.98441e-07,-9.49191e-06,-3.50926e-06,3.98922e-06,7.76319e-07,-9.26289e-07,6.75401e-06,-6.60171e-07,-9.6001e-07,5.99713e-06,-5.95415e-07,3.09606e-07,-1.12361e-06,4.87125e-06,7.41937e-06,
> 
> 这是因为全局均值会对当前均值做平滑操作：mean = momentum * mean + (1.0 - momentum) * saved_mean，batch norm里默认的momentum是0.9， 所以出现你看到的相差10倍的情况

这里似乎看不到momentum * mean对mean的贡献，这是为什么呢？ mean的初始值为0，可是随着迭代进行，mean不应为0呀，也就是说更新后的mean应该大于 0.1*saved_mean ？？

**Comment by ceci3** [comment]
> 储存batchnorm层的参数：batch_norm_0.b_0 batch_norm_0.w_0 batch_norm_0.w_1 batch_norm_0.w_2
> 
> 分别对应了代码里的：bias，scale，mean，variance。 saved_mean的名字是Block_5_layer2_bn1.tmp_0，不是储存的参数。 那么推断时载入的参数？

模型推断是载入的参数是mean，variance，全局均值方差，训练时当前batch计算出来的saved_mean和saved_var是不会保存的

**Comment by ceci3** [comment]
> > > 当batchsize 为10的时候，我把mean_out 和 saved_mean打印出来，发现数值上很接近，但是相差10倍。
> > > 1567660471	mean_out	The place is:CPUPlace
> > > Tensor[Block_5_layer2_bn1.w_1]
> > > shape: [512,]
> > > dtype: f
> > > data: 6.25614e-07,6.05691e-07,3.71493e-07,4.25677e-07,-8.28404e-08,-4.7798e-07,8.96869e-07,2.77358e-07,-9.39467e-07,-4.27785e-07,1.75467e-08,7.98441e-08,-9.49192e-07,-3.50926e-07,3.98922e-07,7.76319e-08,-9.2629e-08,6.75401e-07,-6.60171e-08,-9.6001e-08,5.99713e-07,-5.95415e-08,3.09607e-08,-1.12361e-07,4.87125e-07,7.41937e-07
> > > 1567660471	saved_mean	The place is:CPUPlace
> > > Tensor[Block_5_layer2_bn1.tmp_0]
> > > shape: [512,]
> > > dtype: f
> > > data: 6.25614e-06,6.05691e-06,3.71493e-06,4.25677e-06,-8.28403e-07,-4.7798e-06,8.96869e-06,2.77358e-06,-9.39467e-06,-4.27785e-06,1.75467e-07,7.98441e-07,-9.49191e-06,-3.50926e-06,3.98922e-06,7.76319e-07,-9.26289e-07,6.75401e-06,-6.60171e-07,-9.6001e-07,5.99713e-06,-5.95415e-07,3.09606e-07,-1.12361e-06,4.87125e-06,7.41937e-06,
> > 
> > 
> > 这是因为全局均值会对当前均值做平滑操作：mean = momentum * mean + (1.0 - momentum) * saved_mean，batch norm里默认的momentum是0.9， 所以出现你看到的相差10倍的情况
> 
> 这里似乎看不到momentum * mean对mean的贡献，这是为什么呢？ mean的初始值为0，可是随着迭代进行，mean不应为0呀，也就是说更新后的mean应该大于 0.1*saved_mean ？？

是的，初始的全局mean手动定义为0，之后的计算中  momentum * mean里的mean都是上一个batch输出的全局均值

**Comment by Dely-Yu** [comment]
另外，一个关于 batch_norm 的问题。我想在训练的时候也使用全局mean和variance，这样能够使得训练和测试的行为相同，“当设置use_global_stats为True时，在训练期间也使用全局均值和方差。”

遇到的问题是： 我把 mean打印出来了，发现一直为0，mean不应该总是 0 的。

1567665060	mean	The place is:CPUPlace
Tensor[Block_5_layer0_bn2.w_1]
	shape: [512,]
	dtype: f
	data: 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
1567665060	mean_out	The place is:CPUPlace
Tensor[Block_5_layer0_bn2.w_1]
	shape: [512,]
	dtype: f
	data: 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
1567665060	saved_mean	The place is:CPUPlace
Tensor[Block_5_layer0_bn2.tmp_0]
	shape: [512,]
	dtype: f
	data: 0.00428498,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,0.00428443,
@ceci3

**Comment by ceci3** [comment]
有复现的代码吗？我复现一下看看？

**Comment by Dely-Yu** [comment]
> 有复现的代码吗？我复现一下看看？

代码不是单一个文件，而且代码和数据在一起的才能跑起来。1. 我打个包，发给你？ 2.你登录到我机器上运行试试，那么hi私聊？

**Comment by ceci3** [comment]
use_global_stats设置为True，默认使用的均值和方差就是初始设置的0 和1，而且训练过程中也不会计算新的全局均值方差，所以打印出来的均值都是0
